### ISSUE TITLE
enforce_ownership step of deploy gets slower and slower on each deploy because of complexity in shared/cache/assets

### ISSUE THREAD

**Issue by willkoehler** [ajgon/opsworks_ruby#262]
Over time deployments take longer and longer. When I first build a server, deploys takes ~1 minute. As the server "ages" over several weeks, deploys take longer and longer, eventually > 10 minutes.

In particular the "set user to deploy" step, which runs multiple times during a deploy, starts taking > 1 minute each time it runs. Other's have observer similar phenomena ex: https://forums.aws.amazon.com/thread.jspa?messageID=893671

```
[2021-05-26T18:09:38+00:00] INFO: deploy[shayhi] updated to revision af937d347a24cd884822cdbdaf1abe1e0b04e381
[2021-05-26T18:11:09+00:00] INFO: deploy[shayhi] set user to deploy
[2021-05-26T18:11:09+00:00] INFO: deploy[shayhi] set group to www-data
[2021-05-26T18:11:11+00:00] INFO: deploy[shayhi] copied the cached checkout to /srv/www/shayhi/releases/20210526180933
[2021-05-26T18:12:37+00:00] INFO: deploy[shayhi] set user to deploy
[2021-05-26T18:12:37+00:00] INFO: deploy[shayhi] set group to www-data
[2021-05-26T18:12:37+00:00] INFO: deploy[shayhi] running callback before_migrate
```

I traced this down to the `enforce_ownership` call here: https://github.com/ajgon/opsworks_ruby/blob/3020734f107310410f3b94d582dc0a90f15a21db/libraries/provider_deploy.rb#L269-L275

Basically, the `chown -R` takes longer and longer as the contents of the `/srv/www/[apname]` folder get more complex.

I can confirm this by running `chown` manually on the server

```
ubuntu@rails1:/srv/www/shayhi$ time sudo chown -R deploy:www-data .

real	1m21.160s
user	0m0.187s
sys	0m2.818s
```

Digging deeper, the bulk of this time comes from the shared/cache/assets folder

```
ubuntu@rails1:/srv/www/shayhi$ time sudo chown -R deploy:www-data shared/cache/assets/

real	1m18.517s
user	0m0.097s
sys	0m2.021s
```

I wonder if there's a way to run `chown_R` more surgically so it doesn't need to run through all the files in shared/cache/assets/ on each deploy?

**Comment by inopinatus** [comment]
Yeah, it bugs me too. It's one of those things I file under, "if Chef had been designed better, this wouldn't have been necessary".  In practice, with all my deploys, everything in the target locations ends up owned by deploy or root anyway.

So I drop this into `libraries/no_chown_please.rb`:

```
class Chef::Provider::Deploy
  # ffs, don't slam the block device with a massive and wholly unnecessary recursive chown
  def enforce_ownership
    Chef::Log.info("#{@new_resource} skipping chown of #{@new_resource.deploy_to}")
  end
end
```

**Comment by olbrich** [comment]
Good find.  

I would also encourage you to rebuild your servers on a regular basis anyway to make sure you get the latest OS level security patches.

**Comment by olbrich** [comment]
Could be a useful approach... 
https://unix.stackexchange.com/questions/245274/chown-only-where-needed-speedup-chown

**Comment by willkoehler** [comment]
In my particular case, I was able to solve this by reducing the number of assets (there were ~2000 emoji SVGs in the asset pipeline, which I was able to move somewhere else). It might still be worth solving. But for all my apps, which have ~200 assets, the recursive `chown` is actually fine (runs in 100msec)
